oauthenticator/multiauthenticator.py

```diff
@@ -0,0 +1,124 @@
+"""
+Custom Authenticator to use multiple OAuth providers with JupyterHub
+
+Example of configuration:
+
+    from oauthenticator.github import GitHubOAuthenticator
+    from oauthenticator.google import GoogleOAuthenticator
+
+    c.MultiAuthenticator.authenticators = [
+        (GitHubOAuthenticator, '/github', {
+            'client_id': 'xxxx',
+            'client_secret': 'xxxx',
+            'oauth_callback_url': 'http://example.com/hub/github/oauth_callback'
+        }),
+        (GoogleOAuthenticator, '/google', {
+            'client_id': 'xxxx',
+            'client_secret': 'xxxx',
+            'oauth_callback_url': 'http://example.com/hub/google/oauth_callback'
+        }),
+        (PAMAuthenticator, "/pam", {"service_name": "PAM"}),
+    ]
+
+    c.JupyterHub.authenticator_class = 'oauthenticator.MultiAuthenticator.MultiAuthenticator'
+
+The same Authenticator class can be used several to support different providers.
+
+"""
+from jupyterhub.auth import Authenticator
+from jupyterhub.utils import url_path_join
+from traitlets import List
+
+
+class URLScopeMixin(object):
+    """Mixin class that adds the"""
+
+    scope = ""
+
+    def login_url(self, base_url):
+        return super().login_url(url_path_join(base_url, self.scope))
+
+    def logout_url(self, base_url):
+        return super().logout_url(url_path_join(base_url, self.scope))
+
+    def get_handlers(self, app):
+        handlers = super().get_handlers(app)
+        return [
+            (url_path_join(self.scope, path), handler) for path, handler in handlers
+        ]
+
+
+class MultiAuthenticator(Authenticator):
+    """Wrapper class that allows to use more than one authentication provider
+    for JupyterHub"""
+
+    authenticators = List(help="The subauthenticators to use", config=True)
+
+    def __init__(self, *arg, **kwargs):
+        super().__init__(*arg, **kwargs)
+        self._authenticators = []
+        for (
+            authenticator_klass,
+            url_scope,
+            authenticator_configuration,
+        ) in self.authenticators:
+            configuration = self.trait_values()
+            # Remove this one as it will overwrite the value if the authenticator_klass
+            # makes it configurable and the default value is used (take a look at
+            # GoogleOAuthenticator for example).
+            configuration.pop("login_service")
+
+            class WrapperAuthenticator(URLScopeMixin, authenticator_klass):
+                scope = url_scope
+
```

Comment:
The username conflict might also be managed in the wrapper by adding a prefix to the username.
I managed to make it work with the following code:

```python
username_prefix = service_name+'_'

async def authenticate(self, handler, data=None, **kwargs):
    response = await super().authenticate(handler, data, **kwargs)
    if response is None:
        return None
    elif type(response) == str:
        return self.username_prefix+response
    else:
        response['name'] = self.username_prefix+response['name']
        return response

# def normalize_username(self, username):
#     print("normalize username :",username, super().normalize_username(username))
#     return self.username_prefix+super().normalize_username(username)

def check_allowed(self, username, authentication=None):
    print("check allowed provided :",username)
    return super().check_allowed(username.removeprefix(self.username_prefix), authentication)

def check_blocked_users(self, username, authentication=None):
    print("check blocked provided :",username)
    return super().check_allowed(username.removeprefix(self.username_prefix), authentication)
```

As you can see, I initially tried to add the prefix by overriding `normalize_username` but it appears to be called inside `oauth2`. One drawback is the display of this username's prefix in the upper corner of jupyterhub.